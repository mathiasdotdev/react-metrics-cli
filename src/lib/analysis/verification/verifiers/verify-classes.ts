/**
 * Vérificateur d'utilisation des classes
 */

import { isComment, isInStringWithName } from '@utils/string-utils';
import { Declaration, DeclarationType, isUsed } from '$types/analysis';

/**
 * Vérifie si une utilisation de classe est valide
 */
const isValidUsage = (line: string, name: string): boolean => {
  // Ignorer commentaires
  if (isComment(line)) {
    return false;
  }

  // Ignorer dans strings
  if (isInStringWithName(line, name)) {
    return false;
  }

  // Ignorer les export { nom } car ce n'est pas une vraie utilisation
  const trimmed = line.trim();
  if (
    trimmed.startsWith('export') &&
    trimmed.includes('{') &&
    trimmed.includes(name)
  ) {
    const exportPattern = new RegExp(`export\\s*\\{[^}]*\\b${name}\\b[^}]*\\}`);
    if (exportPattern.test(trimmed)) {
      return false;
    }
  }

  // Patterns d'utilisation de classe
  // Note: We check patterns before isCompleteIdentifier to handle cases where
  // the class name appears as a substring elsewhere in the line

  // new ClassName()
  if (line.includes(`new ${name}`)) {
    return true;
  }

  // extends ClassName
  if (line.includes(`extends ${name}`)) {
    return true;
  }

  // implements ClassName (single or multiple)
  if (line.includes(`implements ${name}`) || line.includes(`, ${name}`)) {
    return true;
  }

  // instanceof ClassName
  if (line.includes(`instanceof ${name}`)) {
    return true;
  }

  // ClassName.staticMethod (static method or property access)
  if (line.includes(`${name}.`)) {
    return true;
  }

  // <ClassName> (generic type)
  if (
    line.includes(`<${name}>`) ||
    line.includes(`<${name},`) ||
    line.includes(`, ${name}>`)
  ) {
    return true;
  }

  // : ClassName (type annotation)
  if (line.includes(`: ${name}`)) {
    return true;
  }

  // (ClassName) or as parameter
  if (
    line.includes(`(${name})`) ||
    line.includes(`(${name},`) ||
    line.includes(`, ${name})`)
  ) {
    return true;
  }

  return false;
};

/**
 * Vérifie si une classe est utilisée
 */
const isClassUsed = (lines: string[], declaration: Declaration): boolean => {
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (!line) continue;

    // Ignorer la ligne de déclaration
    if (i + 1 === declaration.location.line) {
      continue;
    }

    // Vérifier si le nom de la classe apparaît
    if (line.includes(declaration.name)) {
      if (isValidUsage(line, declaration.name)) {
        return true;
      }
    }
  }

  return false;
};

/**
 * Vérifie les utilisations des classes
 */
export const verifyClasses = (
  content: string,
  declarations: Map<string, Declaration>,
): void => {
  const lines = content.split('\n');

  for (const [_key, declaration] of declarations) {
    if (declaration.type !== DeclarationType.CLASS || isUsed(declaration)) {
      continue;
    }

    if (isClassUsed(lines, declaration)) {
      declaration.isUsedLocally = true;
    }
  }
};
