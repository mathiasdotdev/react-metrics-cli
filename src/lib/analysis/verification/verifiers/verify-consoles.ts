/**
 * Vérificateur pour console.* - Tous restent "non utilisés" (code mort)
 */

import { Declaration, DeclarationType } from '$types/analysis';

/**
 * Vérifie les utilisations des console.* (toujours considérés comme code mort)
 */
export const verifyConsoles = (
  _content: string,
  declarations: Map<string, Declaration>,
): void => {
  // Les console.* sont toujours considérés comme du code mort
  // On ne fait rien, ils restent isUsedLocally = false et isImportedExternally = false
  for (const [_key, declaration] of declarations) {
    if (declaration.type === DeclarationType.CONSOLE) {
      declaration.isUsedLocally = false;
      declaration.isImportedExternally = false;
    }
  }
};
