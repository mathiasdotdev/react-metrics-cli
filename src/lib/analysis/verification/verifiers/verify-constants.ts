/**
 * Vérificateur d'utilisation des constantes
 */

import {
  isComment,
  isCompleteIdentifier,
  isInStringWithName,
} from '@utils/string-utils';
import { Declaration, DeclarationType, isUsed } from '$types/analysis';

/**
 * Vérifie si une utilisation de constante est valide
 */
const isValidUsage = (line: string, name: string): boolean => {
  // Ignorer les commentaires
  if (isComment(line)) {
    return false;
  }

  // Ignorer si dans une chaîne de caractères
  if (isInStringWithName(line, name)) {
    return false;
  }

  // Ignorer les export { nom } car ce n'est pas une vraie utilisation
  const trimmed = line.trim();
  if (
    trimmed.startsWith('export') &&
    trimmed.includes('{') &&
    trimmed.includes(name)
  ) {
    const exportPattern = new RegExp(`export\\s*\\{[^}]*\\b${name}\\b[^}]*\\}`);
    if (exportPattern.test(trimmed)) {
      return false;
    }
  }

  // Vérifier que c'est un identifiant complet
  if (!isCompleteIdentifier(line, name)) {
    return false;
  }

  return true;
};

/**
 * Vérifie si une constante est utilisée
 */
const isConstantUsed = (lines: string[], declaration: Declaration): boolean => {
  let usageCount = 0;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (!line) continue;

    // Ignorer la ligne de déclaration
    if (i + 1 === declaration.location.line) {
      continue;
    }

    // Vérifier si le nom de la constante apparaît dans la ligne
    if (line.includes(declaration.name)) {
      if (isValidUsage(line, declaration.name)) {
        usageCount++;
      }
    }
  }

  return usageCount > 0;
};

/**
 * Vérifie les utilisations des constantes
 */
export const verifyConstants = (
  content: string,
  declarations: Map<string, Declaration>,
): void => {
  const lines = content.split('\n');

  for (const [_key, declaration] of declarations) {
    if (declaration.type !== DeclarationType.CONSTANT || isUsed(declaration)) {
      continue;
    }

    if (isConstantUsed(lines, declaration)) {
      declaration.isUsedLocally = true;
    }
  }
};
