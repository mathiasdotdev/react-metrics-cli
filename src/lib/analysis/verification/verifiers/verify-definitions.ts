/**
 * Vérificateur d'utilisation des types et interfaces
 */

import { isComment, isInStringWithName } from '@utils/string-utils';
import { Declaration, DeclarationType, isUsed } from '$types/analysis';

/**
 * Vérifie si une utilisation de définition est valide
 */
const isValidUsage = (line: string, name: string): boolean => {
  // Ignorer commentaires
  if (isComment(line)) {
    return false;
  }

  // Ignorer dans strings
  if (isInStringWithName(line, name)) {
    return false;
  }

  // Ignorer les export { nom } car ce n'est pas une vraie utilisation
  const trimmed = line.trim();
  if (
    trimmed.startsWith('export') &&
    trimmed.includes('{') &&
    trimmed.includes(name)
  ) {
    const exportPattern = new RegExp(`export\\s*\\{[^}]*\\b${name}\\b[^}]*\\}`);
    if (exportPattern.test(trimmed)) {
      return false;
    }
  }

  // Patterns d'utilisation de types
  // Note: We check patterns before isCompleteIdentifier because the name might
  // appear as a substring elsewhere in the line (e.g., "getResult" contains "Result")
  // : TypeName
  if (line.includes(`: ${name}`)) {
    return true;
  }

  // <TypeName>
  if (line.includes(`<${name}>`)) {
    return true;
  }

  // extends TypeName
  if (line.includes(`extends ${name}`)) {
    return true;
  }

  // implements TypeName
  if (line.includes(`implements ${name}`)) {
    return true;
  }

  // as TypeName (type cast)
  if (line.includes(` as ${name}`)) {
    return true;
  }

  // is TypeName (type guard)
  if (line.includes(` is ${name}`)) {
    return true;
  }

  // keyof TypeName
  if (line.includes(`keyof ${name}`)) {
    return true;
  }

  // typeof TypeName
  if (line.includes(`typeof ${name}`)) {
    return true;
  }

  // | TypeName (union type)
  if (line.includes(`| ${name}`)) {
    return true;
  }

  // & TypeName (intersection type)
  if (line.includes(`& ${name}`)) {
    return true;
  }

  // = TypeName (type alias)
  if (line.includes(`= ${name}`)) {
    return true;
  }

  return false;
};

/**
 * Vérifie si une définition est utilisée
 */
const isDefinitionUsed = (
  lines: string[],
  declaration: Declaration,
): boolean => {
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (!line) continue;

    // Ignorer la ligne de déclaration
    if (i + 1 === declaration.location.line) {
      continue;
    }

    // Vérifier si le nom du type apparaît
    if (line.includes(declaration.name)) {
      if (isValidUsage(line, declaration.name)) {
        return true;
      }
    }
  }

  return false;
};

/**
 * Vérifie les utilisations des définitions (types et interfaces)
 */
export const verifyDefinitions = (
  content: string,
  declarations: Map<string, Declaration>,
): void => {
  const lines = content.split('\n');

  for (const [_key, declaration] of declarations) {
    if (
      declaration.type !== DeclarationType.DEFINITION ||
      isUsed(declaration)
    ) {
      continue;
    }

    if (isDefinitionUsed(lines, declaration)) {
      declaration.isUsedLocally = true;
    }
  }
};
