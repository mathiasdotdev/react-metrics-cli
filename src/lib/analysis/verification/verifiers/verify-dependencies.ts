/**
 * Vérificateur d'utilisation des dépendances
 */

import { Declaration, DeclarationType, isUsed } from '$types/analysis';

/**
 * Vérifie si une dépendance est importée dans un fichier
 */
const isImportedInFile = (content: string, packageName: string): boolean => {
  const lines = content.split('\n');

  for (const line of lines) {
    // Ignorer les commentaires
    const trimmed = line.trim();
    if (
      trimmed.startsWith('//') ||
      trimmed.startsWith('/*') ||
      trimmed.startsWith('*')
    ) {
      continue;
    }

    // Vérifier les imports avec 'from'
    if (line.includes('import') && line.includes('from')) {
      // import ... from 'packageName'
      if (
        line.includes(`'${packageName}'`) ||
        line.includes(`"${packageName}"`)
      ) {
        return true;
      }
      // import ... from 'packageName/submodule'
      if (
        line.includes(`'${packageName}/`) ||
        line.includes(`"${packageName}/`)
      ) {
        return true;
      }
    }

    // Side-effect imports sans 'from': import 'dotenv/config'
    if (line.includes('import') && !line.includes('from')) {
      if (
        line.includes(`'${packageName}'`) ||
        line.includes(`"${packageName}"`)
      ) {
        return true;
      }
      if (
        line.includes(`'${packageName}/`) ||
        line.includes(`"${packageName}/`)
      ) {
        return true;
      }
    }

    // export ... from 'packageName'
    if (line.includes('export') && line.includes('from')) {
      if (
        line.includes(`'${packageName}'`) ||
        line.includes(`"${packageName}"`)
      ) {
        return true;
      }
      if (
        line.includes(`'${packageName}/`) ||
        line.includes(`"${packageName}/`)
      ) {
        return true;
      }
    }

    // require('packageName')
    if (
      line.includes('require') &&
      (line.includes(`'${packageName}'`) || line.includes(`"${packageName}"`))
    ) {
      return true;
    }

    // Dynamic import: import('packageName')
    if (
      line.includes('import(') &&
      (line.includes(`'${packageName}'`) || line.includes(`"${packageName}"`))
    ) {
      return true;
    }
  }

  return false;
};

/**
 * Vérifie si une dépendance est utilisée
 */
const isDependencyUsed = (
  declaration: Declaration,
  fileContents: Map<string, string>,
): boolean => {
  // Parcourir tous les fichiers pour voir si la dépendance est importée
  for (const [_filePath, content] of fileContents) {
    if (isImportedInFile(content, declaration.name)) {
      return true;
    }
  }

  return false;
};

/**
 * Vérifie les utilisations des dépendances
 */
export const verifyDependencies = (
  _content: string,
  declarations: Map<string, Declaration>,
  fileContents: Map<string, string>,
): void => {
  // Vérifier si la dépendance apparaît dans les imports de tous les fichiers
  for (const [_key, declaration] of declarations) {
    if (
      declaration.type !== DeclarationType.DEPENDENCY ||
      isUsed(declaration)
    ) {
      continue;
    }

    if (isDependencyUsed(declaration, fileContents)) {
      declaration.isUsedLocally = true;
    }
  }
};
