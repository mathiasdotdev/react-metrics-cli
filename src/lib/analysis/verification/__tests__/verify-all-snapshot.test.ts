import { describe, expect, test } from 'bun:test';
import {
  createDeclaration,
  runVerificationSnapshot,
  runVerificationSnapshotWithFiles,
} from './helpers/verificationSnapshotHelper';
import { DeclarationType } from '$types/analysis';
import { VERIFICATION_SAMPLES } from '$tests/fixtures/code-samples';

// Import verifiers
import { verifyFunctions } from '../verifiers/verify-functions';
import { verifyClasses } from '../verifiers/verify-classes';
import { verifyConstants } from '../verifiers/verify-constants';
import { verifyProps } from '../verifiers/verify-props';
import { verifyDefinitions } from '../verifiers/verify-definitions';
import {
  verifyLocalUsage,
  verifyExternalImports,
} from '../verifiers/verify-exports';
import { verifyConsoles } from '../verifiers/verify-consoles';
import { verifyDependencies } from '../verifiers/verify-dependencies';
import { verifyDeprecated } from '../verifiers/verify-deprecated';
import type { Declaration } from '$types/analysis';

// Wrapper pour verifyExports qui combine les deux vérificateurs
const verifyExports = (
  fileContents: Map<string, string>,
  declarations: Map<string, Declaration>,
) => {
  // Les deux fonctions prennent 3 paramètres : (_content, declarations, fileContents)
  verifyLocalUsage('', declarations, fileContents);
  verifyExternalImports('', declarations, fileContents);
};

describe('Verification - All Verifiers Snapshot Tests', () => {
  test('verifyFunctions: all cases', () => {
    const testCases = {
      functionCalled: {
        declarations: [
          createDeclaration('myFunction', DeclarationType.FUNCTION),
        ],
        content: VERIFICATION_SAMPLES.functionCalled,
      },
      functionAsCallback: {
        declarations: [
          createDeclaration('handleClick', DeclarationType.FUNCTION),
        ],
        content: VERIFICATION_SAMPLES.functionAsCallback,
      },
      functionInComment: {
        declarations: [
          createDeclaration('myFunction', DeclarationType.FUNCTION),
        ],
        content: VERIFICATION_SAMPLES.functionInComment,
      },
      functionInString: {
        declarations: [
          createDeclaration('myFunction', DeclarationType.FUNCTION),
        ],
        content: VERIFICATION_SAMPLES.functionInString,
      },
      functionInJSX: {
        declarations: [
          createDeclaration(
            'handleSubmit',
            DeclarationType.FUNCTION,
            '/test/file.tsx',
          ),
        ],
        content: VERIFICATION_SAMPLES.functionInJSX,
      },
    };

    const result = runVerificationSnapshot(verifyFunctions, testCases);
    expect(result).toMatchSnapshot();
  });

  test('verifyClasses: all cases', () => {
    const testCases = {
      classInstantiated: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: VERIFICATION_SAMPLES.classInstantiated,
      },
      classExtended: {
        declarations: [createDeclaration('BaseClass', DeclarationType.CLASS)],
        content: VERIFICATION_SAMPLES.classExtended,
      },
      classInInstanceof: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: VERIFICATION_SAMPLES.classInInstanceof,
      },
      classAsType: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: VERIFICATION_SAMPLES.classAsType,
      },
      classInComment: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: '// MyClass is deprecated\nconst x = 1;',
      },
      classInString: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: 'const msg = "MyClass should not be used";',
      },
      classInExportBraces: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: 'export { MyClass };',
      },
      classWithImplementsComma: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: 'class Test implements IFirst, MyClass { }',
      },
      classStaticAccess: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: 'const x = MyClass.staticMethod();',
      },
      classInGenericWithComma: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: 'const map: Map<string, MyClass> = new Map();',
      },
      classAsParameter: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: 'function test(first, MyClass) { }',
      },
      classAsFirstParameter: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: 'function test(MyClass, second) { }',
      },
      classAsLastParameter: {
        declarations: [createDeclaration('MyClass', DeclarationType.CLASS)],
        content: 'function test(first, second, MyClass) { }',
      },
    };

    const result = runVerificationSnapshot(verifyClasses, testCases);
    expect(result).toMatchSnapshot();
  });

  test('verifyConstants: all cases', () => {
    const testCases = {
      constantReferenced: {
        declarations: [createDeclaration('API_URL', DeclarationType.CONSTANT)],
        content: VERIFICATION_SAMPLES.constantReferenced,
      },
      constantInExpression: {
        declarations: [
          createDeclaration('MAX_COUNT', DeclarationType.CONSTANT),
        ],
        content: VERIFICATION_SAMPLES.constantInExpression,
      },
      constantInTemplateInterpolation: {
        declarations: [createDeclaration('USERNAME', DeclarationType.CONSTANT)],
        content: VERIFICATION_SAMPLES.constantInTemplateInterpolation,
      },
      constantInSpread: {
        declarations: [
          createDeclaration('BASE_PROPS', DeclarationType.CONSTANT),
        ],
        content: VERIFICATION_SAMPLES.constantInSpread,
      },
    };

    const result = runVerificationSnapshot(verifyConstants, testCases);
    expect(result).toMatchSnapshot();
  });

  test('verifyProps: all cases', () => {
    const testCases = {
      propAccessed: {
        declarations: [createDeclaration('name', DeclarationType.PROP)],
        content:
          'const name: string = "John";\nconst greeting = `Hello ${name}`;',
      },
      propInDestructuring: {
        declarations: [createDeclaration('email', DeclarationType.PROP)],
        content:
          'const email: string = "test@test.com";\nconst { email } = user;',
      },
      propInSpread: {
        declarations: [createDeclaration('title', DeclarationType.PROP)],
        content: 'const title: string = "Test";\nconst props = { ...title };',
      },
    };

    const result = runVerificationSnapshot(verifyProps, testCases);
    expect(result).toMatchSnapshot();
  });

  test('verifyDefinitions: all cases', () => {
    const testCases = {
      typeUsedInAnnotation: {
        declarations: [createDeclaration('User', DeclarationType.DEFINITION)],
        content: VERIFICATION_SAMPLES.typeUsedInAnnotation,
      },
      interfaceImplemented: {
        declarations: [
          createDeclaration('IService', DeclarationType.DEFINITION),
        ],
        content: VERIFICATION_SAMPLES.interfaceImplemented,
      },
      typeInGeneric: {
        declarations: [createDeclaration('Result', DeclarationType.DEFINITION)],
        content: VERIFICATION_SAMPLES.typeInGeneric,
      },
      typeInUnion: {
        declarations: [createDeclaration('Status', DeclarationType.DEFINITION)],
        content: VERIFICATION_SAMPLES.typeInUnion,
      },
      definitionInString: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'const msg = "UserType should be used";',
      },
      definitionInComment: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: '// UserType is deprecated',
      },
      definitionInExport: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'export { UserType };',
      },
      definitionNotInExtends: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'interface Other extends BaseType { }',
      },
      definitionNotInImplements: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'class Test implements IBase { }',
      },
      definitionNotInAs: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'const x = value as BaseType;',
      },
      definitionNotInIs: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'function check(x): x is BaseType { return true; }',
      },
      definitionNotInKeyof: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'type Keys = keyof BaseType;',
      },
      definitionNotInTypeof: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'type T = typeof BaseType;',
      },
      definitionNotInIntersection: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'type Combined = BaseType & OtherType;',
      },
      definitionInIntersection: {
        declarations: [
          createDeclaration('UserType', DeclarationType.DEFINITION),
        ],
        content: 'type Combined = BaseType & UserType;',
      },
    };

    const result = runVerificationSnapshot(verifyDefinitions, testCases);
    expect(result).toMatchSnapshot();
  });

  test('verifyExports: all cases', () => {
    const testCases = {
      exportImportedInAnotherFile: {
        declarations: [
          createDeclaration(
            'myFunction',
            DeclarationType.EXPORT,
            '/src/utils.ts',
          ),
        ],
        fileContents: new Map([
          ['/src/utils.ts', 'export function myFunction() {}'],
          ['/src/app.ts', VERIFICATION_SAMPLES.exportImported],
        ]),
      },
      exportWithMultipleImports: {
        declarations: [
          createDeclaration('myExport', DeclarationType.EXPORT, '/src/lib.ts'),
        ],
        fileContents: new Map([
          ['/src/lib.ts', 'export const myExport = true'],
          ['/src/main.ts', VERIFICATION_SAMPLES.exportWithMultipleImports],
        ]),
      },
      exportReExported: {
        declarations: [
          createDeclaration(
            'myFunction',
            DeclarationType.EXPORT,
            '/src/utils.ts',
          ),
        ],
        fileContents: new Map([
          ['/src/utils.ts', 'export function myFunction() {}'],
          ['/src/index.ts', VERIFICATION_SAMPLES.exportReExported],
        ]),
      },
      importWithoutFrom: {
        declarations: [
          createDeclaration(
            'myExport',
            DeclarationType.EXPORT,
            '/src/file1.ts',
          ),
        ],
        fileContents: new Map([
          ['/src/file1.ts', 'export const myExport = 1;'],
          ['/src/file2.ts', 'import myExport;'], // malformed import
        ]),
      },
      importWithRelativeParentPath: {
        declarations: [
          createDeclaration(
            'helper',
            DeclarationType.EXPORT,
            '/src/utils/file1.ts',
          ),
        ],
        fileContents: new Map([
          ['/src/utils/file1.ts', 'export const helper = 1;'],
          [
            '/src/components/deep/file2.ts',
            'import { helper } from "../../utils/file1";',
          ],
        ]),
      },
      importWithComplexRelativePath: {
        declarations: [
          createDeclaration(
            'util',
            DeclarationType.EXPORT,
            '/src/lib/utils/util.ts',
          ),
        ],
        fileContents: new Map([
          ['/src/lib/utils/util.ts', 'export const util = 1;'],
          [
            '/src/features/user/components/form.tsx',
            'import { util } from "../../../lib/utils/util";',
          ],
        ]),
      },
    };

    const result = runVerificationSnapshotWithFiles(verifyExports, testCases);
    expect(result).toMatchSnapshot();
  });

  test('verifyConsoles: all cases', () => {
    const testCases = {
      consoleLog: {
        declarations: [createDeclaration('console', DeclarationType.CONSOLE)],
        content: 'console.log("test")',
      },
      consoleError: {
        declarations: [createDeclaration('console', DeclarationType.CONSOLE)],
        content: 'console.error("error")',
      },
      consoleWarn: {
        declarations: [createDeclaration('console', DeclarationType.CONSOLE)],
        content: 'console.warn("warning")',
      },
    };

    const result = runVerificationSnapshot(verifyConsoles, testCases);
    expect(result).toMatchSnapshot();
  });

  test('verifyDependencies: all cases', () => {
    const testCases = {
      dependencyImported: {
        declarations: [
          createDeclaration(
            'react',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/file.tsx', VERIFICATION_SAMPLES.dependencyImported],
        ]),
      },
      dependencyRequired: {
        declarations: [
          createDeclaration(
            'express',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/server.js', VERIFICATION_SAMPLES.dependencyRequired],
        ]),
      },
      dependencyScopedImport: {
        declarations: [
          createDeclaration(
            '@types/node',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/types.ts', VERIFICATION_SAMPLES.dependencyScopedImport],
        ]),
      },
      dependencySubmodule: {
        declarations: [
          createDeclaration(
            'fs',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/utils.ts', VERIFICATION_SAMPLES.dependencySubmodule],
        ]),
      },
      sideEffectImportSingleQuote: {
        declarations: [
          createDeclaration(
            'dotenv',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([['/test/config.ts', "import 'dotenv/config';"]]),
      },
      sideEffectImportWithSubmodule: {
        declarations: [
          createDeclaration(
            'react',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/profiler.ts', 'import "react/profiler";'],
        ]),
      },
      reExportFromPackage: {
        declarations: [
          createDeclaration(
            'react',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/index.ts', 'export { Component } from "react";'],
        ]),
      },
      reExportWithSubmodule: {
        declarations: [
          createDeclaration(
            'lodash',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/utils.ts', 'export * from "lodash/merge";'],
        ]),
      },
      dynamicImport: {
        declarations: [
          createDeclaration(
            'chalk',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/logger.ts', 'const mod = await import("chalk");'],
        ]),
      },
      requireWithSubmodule: {
        declarations: [
          createDeclaration(
            'fs',
            DeclarationType.DEPENDENCY,
            '/test/package.json',
          ),
        ],
        fileContents: new Map([
          ['/test/file.js', 'const promises = require("fs/promises");'],
        ]),
      },
    };

    const result = runVerificationSnapshotWithFiles(
      verifyDependencies,
      testCases,
    );
    expect(result).toMatchSnapshot();
  });

  test('verifyDeprecated: all cases', () => {
    const testCases = {
      deprecatedDeclaration: {
        declarations: [
          createDeclaration('oldFunction', DeclarationType.FUNCTION),
        ],
        content: '/** @deprecated */\nfunction oldFunction() {}',
      },
      nonDeprecatedDeclaration: {
        declarations: [
          createDeclaration('newFunction', DeclarationType.FUNCTION),
        ],
        content: 'function newFunction() {}',
      },
    };

    const result = runVerificationSnapshot(verifyDeprecated, testCases);
    expect(result).toMatchSnapshot();
  });
});
