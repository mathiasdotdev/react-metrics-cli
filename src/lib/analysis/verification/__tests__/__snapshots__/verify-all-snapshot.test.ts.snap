// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Verification - All Verifiers Snapshot Tests verifyFunctions: all cases 1`] = `
"
=== functionAsCallback ===
FUNCTION | handleClick | used: true

=== functionCalled ===
FUNCTION | myFunction | used: true

=== functionInComment ===
FUNCTION | myFunction | used: false

=== functionInJSX ===
FUNCTION | handleSubmit | used: true

=== functionInString ===
FUNCTION | myFunction | used: false"
`;

exports[`Verification - All Verifiers Snapshot Tests verifyClasses: all cases 1`] = `
"
=== classAsFirstParameter ===
CLASS | MyClass | used: false

=== classAsLastParameter ===
CLASS | MyClass | used: false

=== classAsParameter ===
CLASS | MyClass | used: false

=== classAsType ===
CLASS | MyClass | used: true

=== classExtended ===
CLASS | BaseClass | used: true

=== classInComment ===
CLASS | MyClass | used: false

=== classInExportBraces ===
CLASS | MyClass | used: false

=== classInGenericWithComma ===
CLASS | MyClass | used: false

=== classInInstanceof ===
CLASS | MyClass | used: true

=== classInString ===
CLASS | MyClass | used: false

=== classInstantiated ===
CLASS | MyClass | used: true

=== classStaticAccess ===
CLASS | MyClass | used: false

=== classWithImplementsComma ===
CLASS | MyClass | used: false"
`;

exports[`Verification - All Verifiers Snapshot Tests verifyConstants: all cases 1`] = `
"
=== constantInExpression ===
CONSTANT | MAX_COUNT | used: true

=== constantInSpread ===
CONSTANT | BASE_PROPS | used: true

=== constantInTemplateInterpolation ===
CONSTANT | USERNAME | used: true

=== constantReferenced ===
CONSTANT | API_URL | used: true"
`;

exports[`Verification - All Verifiers Snapshot Tests verifyProps: all cases 1`] = `
"
=== propAccessed ===
PROP | name | used: true

=== propInDestructuring ===
PROP | email | used: true

=== propInSpread ===
PROP | title | used: true"
`;

exports[`Verification - All Verifiers Snapshot Tests verifyDefinitions: all cases 1`] = `
"
=== definitionInComment ===
DEFINITION | UserType | used: false

=== definitionInExport ===
DEFINITION | UserType | used: false

=== definitionInIntersection ===
DEFINITION | UserType | used: false

=== definitionInString ===
DEFINITION | UserType | used: false

=== definitionNotInAs ===
DEFINITION | UserType | used: false

=== definitionNotInExtends ===
DEFINITION | UserType | used: false

=== definitionNotInImplements ===
DEFINITION | UserType | used: false

=== definitionNotInIntersection ===
DEFINITION | UserType | used: false

=== definitionNotInIs ===
DEFINITION | UserType | used: false

=== definitionNotInKeyof ===
DEFINITION | UserType | used: false

=== definitionNotInTypeof ===
DEFINITION | UserType | used: false

=== interfaceImplemented ===
DEFINITION | IService | used: true

=== typeInGeneric ===
DEFINITION | Result | used: true

=== typeInUnion ===
DEFINITION | Status | used: true

=== typeUsedInAnnotation ===
DEFINITION | User | used: true"
`;

exports[`Verification - All Verifiers Snapshot Tests verifyExports: all cases 1`] = `
"
=== exportImportedInAnotherFile ===
EXPORT | myFunction | used: true

=== exportReExported ===
EXPORT | myFunction | used: true

=== exportWithMultipleImports ===
EXPORT | myExport | used: true

=== importWithComplexRelativePath ===
EXPORT | util | used: true

=== importWithRelativeParentPath ===
EXPORT | helper | used: true

=== importWithoutFrom ===
EXPORT | myExport | used: false"
`;

exports[`Verification - All Verifiers Snapshot Tests verifyConsoles: all cases 1`] = `
"
=== consoleError ===
CONSOLE | console | used: false

=== consoleLog ===
CONSOLE | console | used: false

=== consoleWarn ===
CONSOLE | console | used: false"
`;

exports[`Verification - All Verifiers Snapshot Tests verifyDependencies: all cases 1`] = `
"
=== dependencyImported ===
DEPENDENCY | react | used: true

=== dependencyRequired ===
DEPENDENCY | express | used: true

=== dependencyScopedImport ===
DEPENDENCY | @types/node | used: true

=== dependencySubmodule ===
DEPENDENCY | fs | used: true

=== dynamicImport ===
DEPENDENCY | chalk | used: true

=== reExportFromPackage ===
DEPENDENCY | react | used: true

=== reExportWithSubmodule ===
DEPENDENCY | lodash | used: true

=== requireWithSubmodule ===
DEPENDENCY | fs | used: false

=== sideEffectImportSingleQuote ===
DEPENDENCY | dotenv | used: true

=== sideEffectImportWithSubmodule ===
DEPENDENCY | react | used: true"
`;

exports[`Verification - All Verifiers Snapshot Tests verifyDeprecated: all cases 1`] = `
"
=== deprecatedDeclaration ===
FUNCTION | oldFunction | used: false

=== nonDeprecatedDeclaration ===
FUNCTION | newFunction | used: false"
`;
